### ISSUE TITLE
Bauer - Sandwich attack in `reinvestReward()` Function

### ISSUE THREAD

**Issue by sherlock-admin** [sherlock-audit/2023-10-notional-judging#20]
Bauer

high

# Sandwich attack in `reinvestReward()` Function

## Summary
The `reinvestReward()` function is susceptible to a sandwich attack, where malicious users can exploit the lack of additional vault share minting to prematurely withdraw rewards. This occurs by executing `depositFromNotional()` before `reinvestReward()`, acquiring LP tokens, and later using `redeemFromNotional()` to exchange LP tokens for tokens. A bad actor can receive rewards without the need for a specified waiting period after depositing funds.

## Vulnerability Detail
The calculation for vaultShares in the `SingleSidedLPVaultBase._mintVaultShares()` is determined by multiplying the number of LP tokens (lpTokens) acquired by the global total of vault shares (`state.totalVaultSharesGlobal`). 
```solidity

    function _mintVaultShares(uint256 lpTokens) internal returns (uint256 vaultShares) {
        StrategyVaultState memory state = VaultStorage.getStrategyVaultState();
        if (state.totalPoolClaim == 0) {
            // Vault Shares are in 8 decimal precision
            vaultShares = (lpTokens * uint256(Constants.INTERNAL_TOKEN_PRECISION)) / POOL_PRECISION();
        } else {
            vaultShares = (lpTokens * state.totalVaultSharesGlobal) / state.totalPoolClaim;
        }

```
The calculation for poolClaim in the `SingleSidedLPVaultBase._redeemVaultShares` function is determined by multiplying the number of vaultShares by the total value of assets in the pool (`state.totalPoolClaim`), and then dividing the result by the global total of vault shares (`state.totalVaultSharesGlobal`).
```solidity
    function _redeemVaultShares(uint256 vaultShares) internal returns (uint256 poolClaim) {
        StrategyVaultState memory state = VaultStorage.getStrategyVaultState();
        // Will revert on divide by zero, which is the correct behavior
        poolClaim = (vaultShares * state.totalPoolClaim) / state.totalVaultSharesGlobal;

        state.totalPoolClaim -= poolClaim;
        // Will revert on underflow if vault shares is greater than total shares global
        state.totalVaultSharesGlobal -= vaultShares.toUint80();
        state.setStrategyVaultState();
    }

```

This `SingleSidedLPVaultBase.reinvestReward()` function is designed to handle the reinvestment of rewards obtained from a strategy. 
In this function, where the LP token amount is increased without minting additional vault shares, there is a vulnerability to sandwich attack. 
```solidity
  poolClaimAmount = _joinPoolAndStake(amounts, minPoolClaim);

        // Increase LP token amount without minting additional vault shares
        StrategyVaultState memory state = VaultStorage.getStrategyVaultState();
        state.totalPoolClaim += poolClaimAmount;
        state.setStrategyVaultState();

```
Malicious actors can monitor transactions within the trading pool, execute a deposit operation using `depositFromNotional() `before the `reinvestReward()` function is called, acquire LP tokens, and subsequently invoke `redeemFromNotional()` after the execution of `reinvestReward()` to exchange LP tokens for tokens. The issue arises because `reinvestReward()` only updates the LP token amount without minting additional vault shares. This allows users to exploit the sandwich attack, enabling them to promptly withdraw the corresponding rewards gained from the attack.

## Impact
It leads to unfair advantages and financial losses for other users within the trading pool.
## Code Snippet
https://github.com/sherlock-audit/2023-10-notional/blob/main/leveraged-vaults/contracts/vaults/common/SingleSidedLPVaultBase.sol#L385-L411
## Tool used

Manual Review

## Recommendation
The recommendation is to allow users to redeem their funds only after a specified waiting period following the deposit

Duplicate of #48
